lib/puppet/pops/loader/task_instantiator.rb

```diff
@@ -3,66 +3,95 @@
-      if source_ref.end_with?('.json')
-        metadata = source_ref
-      elsif task_source.nil?
-        task_source = source_ref
-      else
-        raise ArgumentError, _('Only one file can exists besides the .json file for task %{name} in directory %{directory}') %
-          { name: name, directory: File.dirname(source_refs[0]) }
-      end
-    end
-
-    if task_source.nil?
-      raise ArgumentError, _('No source besides task metadata was found in directory %{directory} for task %{name}') %
-        { name: name, directory: File.dirname(source_refs[0]) }
-    end
-    create_task(loader, name, task_source, metadata)
-  end
-
-  def self.create_task(loader, name, task_source, metadata)
+  def self.load_metadata(loader, metadata)
     if metadata.nil?
-      create_task_from_hash(name, task_source, EMPTY_HASH)
+      EMPTY_HASH
     else
       json_text = loader.get_contents(metadata)
       begin
-        create_task_from_hash(name, task_source, Puppet::Util::Json.load(json_text) || EMPTY_HASH)
+        Puppet::Util::Json.load(json_text).freeze || EMPTY_HASH
       rescue Puppet::Util::Json::ParseError => ex
         raise Puppet::ParseError.new(ex.message, metadata)
-      rescue Types::TypeAssertionError => ex
-        # Not strictly a parser error but from the users perspective, the file content didn't parse properly. The
-        # ParserError also conveys file info (even though line is unknown)
-        msg = _('Failed to load metadata for task %{name}: %{reason}') % { :name => name, :reason => ex.message }
-        raise Puppet::ParseError.new(msg, metadata)
       end
     end
   end
 
-  def self.create_task_from_hash(name, task_source, hash)
+  def self.validate_implementations(typed_name, directory, metadata, executables)
+    name = typed_name.name
+    basename = typed_name.name_parts[1] || 'init'
+    # If 'implementations' is defined, it needs to mention at least one
+    # implementation, and everything it mentions must exist.
+    if metadata.key?('implementations')
+      if metadata['implementations'].is_a?(Array)
+        metadata['implementations'].map do |impl|
+          path = executables.find { |real_impl| File.basename(real_impl) == impl['name'] }
+          if path
+            { "name" => impl['name'], "requirements" => impl.fetch('requirements', []), "path" => path }
+          else
+            raise ArgumentError, _("Task metadata for task %{name} specifies missing implementation %{implementation}") %
+              { name: name, implementation: impl['name'] }
+          end
+        end
+      else
+        # If 'implementations' is the wrong type, we just pass it through and
+        # let the task type definition reject it.
+        metadata['implementations']
+      end
+    # If implementations isn't defined, then we use executables matching the
+    # task name, and only one may exist.
+    else
+      implementations = executables.select { |impl| File.basename(impl, '.*') == basename }
```

Comment:
We don't use `basename` until here, why define it so early in the function?

Reply:
It makes sense to me to define vars at the top of a function. But I'm not sure what common practice in puppet repos is (shrug)

Reply:
I think this is an artifact of having extracted `validate_implementations` from `create` in the middle of my work. I can move it down if you like, but I don't have an opinion either way.

Reply:
🤷‍♂️